### ISSUE TITLE
CKEditor - Add Enhanced Image Plugin

### ISSUE THREAD

**Issue by melloware** [primefaces-extensions/primefaces-extensions.github.com#435]
@jxmai looks like this has been a popular plugin request i wonder if we should just add it to the CKEditor by default?

Enhanced Image Plugin = http://ckeditor.com/addon/image2

**Requests:**
http://forum.primefaces.org/viewtopic.php?f=14&t=48571

http://forum.primefaces.org/viewtopic.php?f=14&t=48585

**Comment by tandraschko** [comment]
only +1 if you we create a java class like resourcemodifier to update everything automatically
It's to time intensive if we also must add plugins manually each time

**Comment by tandraschko** [comment]
And I don't think it's very easy to integrate a image upload, it's requires some hacking of the plugin.

**Comment by jxmai** [comment]
I like the idea of uploading image from the local machine, but there are **two main challenges** after a quick glance.

1. The `enhanced image plugin` has other plugin dependencies  that we don't have at the moment.

For example, we have` dialog` but we don't have `widget`. Open `widget`, we only have `clipboard` but missing `Line Utilities` and `Widget Selection`... and every plugin might need potential hack in order to make it work.

![snapshot3](https://cloud.githubusercontent.com/assets/10467831/21614007/686174e2-d19d-11e6-8023-804b9d06f13a.png)

![snapshot3](https://cloud.githubusercontent.com/assets/10467831/21614108/c6dc8ae8-d19d-11e6-8a25-71d03abe3607.png)

2. This feature might also involve ckeditor core code changes since we need to store the image somewhere on the server. 

Take github for example, dragging the image generating a `URL + hash`. (If I understand it correctly. The image has been stored in the DB already, and the hash is just a reference). If The DB transaction happens on drag, not sure how this task can be done using JSF. (We need to link hash to image outputStream that has been stored in Database)

This task is different from imageUpload in primefaces since we cannot define a `UploadedFile` in the bean (The image should be embedded within the text). 

---

Any thought ?

**Comment by melloware** [comment]
You might be right.  Might not be worth the trouble.  Want me to close this as WontFix?

**Comment by jxmai** [comment]
Yes, I think this task needs to be done combining with database so it is **out of scope**.  

---

My workaround is combining pe:ckeditor with p:fileUpload

```xhtml
<composite:interface>

	</composite:interface>

	<composite:implementation>
		<pe:ckEditor id="editor"
				 height="700" 
				 width="1000"
				 value="#{editorBean.text}"/>
		
		<p:fileUpload id="upload" 
	               fileUploadListener="#{editorBean.uploadListener}"
	               label="Insert image"
	               mode="advanced"
	               allowTypes="/(\.|\/)(gif|jpe?g|png)$/"
	               update="editor">
	 	 </p:fileUpload>
	</composite:implementation>
```

There are three custom steps follows:
1. on `uploadListener`, send the file to the database and GET a hash back
2. re-constructor the text and put the URL with hash in.
3. configure HTTP servlet to map the image URL into outputStream.

In the end, we can achieve github like image upload....

**Comment by melloware** [comment]
Understood.

**Comment by krishnakittu** [comment]
HI @jxmai . I tried to use the same code. I generated hash and tries to append to the already existing content in editor. The editor content comes as null in the listener method. How to solve this issue

**Comment by jxmai** [comment]
> The editor content comes as null in the listener method. How to solve this issue

Hi @krishnakittu, there shouldn't be such problem. Please double-check if you wrap components within a form before invoking form submit. I can give you some test sample for you to check.

```html
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml"
      xmlns:ui="http://java.sun.com/jsf/facelets"
      xmlns:f="http://java.sun.com/jsf/core"
      xmlns:p="http://primefaces.org/ui"
      xmlns:pe="http://primefaces.org/ui/extensions"
      xmlns:h="http://java.sun.com/jsf/html">

    <h:head>   
        <title>PrimeFaces Test</title>
    </h:head>
    <h:body>
        <h:form>
	        <h1>#{testView.testString}</h1>
	        <pe:ckEditor id="editor"
					 value="#{testView.testString}"/>
					 
			<p:fileUpload id="upload" 
		               fileUploadListener="#{testView.uploadListener}"
		               label="Insert image"
		               mode="advanced"
		               allowTypes="/(\.|\/)(gif|jpe?g|png)$/"
		               update="editor">
		 	 </p:fileUpload>
        </h:form>
    </h:body>
</html>
```

```java
package org.primefaces.test;

import java.io.Serializable;

import javax.annotation.PostConstruct;
import javax.faces.application.FacesMessage;
import javax.faces.bean.ManagedBean;
import javax.faces.bean.ViewScoped;
import javax.faces.context.FacesContext;

import org.primefaces.event.FileUploadEvent;

@ManagedBean(name = "testView")
@ViewScoped
public class TestView implements Serializable {

    /**
     * 
     */
    private static final long serialVersionUID = 7705670082140081684L;

    private String testString;

    private String fakeHash = "abcde";

    @PostConstruct
    public void init() {
        testString = "Welcome to PrimeFaces!!!";
    }

    public void uploadListener(FileUploadEvent event) {
        FacesMessage msg = new FacesMessage("Succesful",
                event.getFile().getFileName() + " is uploaded.");
        FacesContext.getCurrentInstance().addMessage(null, msg);

        // skip some steps, just assume we have the hash back

        testString = testString.concat(fakeHash);
    }

    public String getTestString() {
        return testString;
    }

    public void setTestString(String testString) {
        this.testString = testString;
    }

    public String getFakeHash() {
        return fakeHash;
    }
}

```

**Comment by krishnakittu** [comment]
HI @jxmai . Thanks for sharing the code. I also did the same thing. But here the problem is if you enter any text in CKeditor before file upload the same text is erased after upload of image. Can you please help me out about this.

**Comment by jxmai** [comment]
@krishnakittu OK, I see. The problem here is that when you if you edit the editor before invoking fileUpload. The content within the editor is wrapped within a `p` tag

For instance, `Welcome to PrimeFaces!!!` becomes `<p>Welcome to PrimeFaces!!!</p>` before calling string concatenation. You don't want to left the hash outside of the p tag.

I would suggest you to extract the content using some custom code or using some library such as jsoup https://jsoup.org/ before manipulating the string.

**Comment by jxmai** [comment]
@krishnakittu OK, I added some workaround to the sample https://github.com/jxmai/Primefaces-Test/tree/PFE_435. You could give it a try.

**Comment by krishnakittu** [comment]
Million Thanks @jxmai  for your support. 👍

**Comment by jxmai** [comment]
Great ! No problem :smiley:

**Comment by krishnakittu** [comment]
Hi Team,
Is there any way to place fileupload button of primefaces as one of the CKEditor Button.

**Comment by cnsgithub** [comment]
I just stumbled upon this issue and wondered if integrating image upoad in CKEditor would really be that difficult.

Since CKEditor actually submits HTML to the server, why not just integrate an additional image upload button to CKEditor (don't know how easy this would be?), use javascript's [FileReader.readAsDataURL](https://developer.mozilla.org/en-US/docs/Web/API/FileReader/readAsDataURL) API to get the image content as Base64 encoded data URL, then just include `<img src="data:image/png;base64," />` in the editor's HTML and that's it. And don't forget to tell the backend side HTML sanitizer to allow images containing `data` URLs only.

Example:
````html
<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABsAAAAUCAYAAAB8gkaAAAAAZElEQVR4Xu3NsQ2AMAxE0T8uCDFcZqBnA3QbwAgmkVy5jlNEKd41V3wAVTaA8IlHBuETjwzCJx4ZhE88MgifeGQQG5QDrmytw1OLlQ2gFeth5th3Ut6dK1vrYDeqbACtWA8Tx34Nt+ewIgKvcwAAAABJRU5ErkJggg==" />
````
http://jsfiddle.net/dg5rf160/

**Comment by krishnakittu** [comment]
HI Team,

With the CK Editor 4 We can easily integrate the upload image feature. Link [here](https://ckeditor.com/docs/ckeditor4/latest/examples/fileupload.html)

The idea was we can add one more property **filebrowserUploadUrl** to the CKeditor component.  Against this property we will give POST URL which will handle the File upload capabilities.

 wb.attr("filebrowserUploadUrl", ckEditor.getFilebrowserUploadUrl());

If you are ok with the thoughts i will provide a PR for it.

**Comment by melloware** [comment]
Sure thing submit a PR!
